File: src/lib/solver/seeds.ts
export const MORPHOLOGIES = [
  {
    id: "protoplast",
    label: "Protoplast",
    blurb: "Spherical cell. Uniform turgor, incompressible interior.",
  },
  {
    id: "budding",
    label: "Budding",
    blurb: "Localized turgor on one flank. Yeast-like outgrowth.",
  },
  {
    id: "fission",
    label: "Fission",
    blurb: "Equatorial constriction, polar expansion. Cytokinesis.",
  },
  {
    id: "polar",
    label: "Polar",
    blurb: "Tip growth. Turgor concentrated on the +Y cap.",
  },
  {
    id: "blastula",
    label: "Blastula",
    blurb: "Hollow shell. Cortex under turgor, empty lumen.",
  },
  {
    id: "dual",
    label: "Dual",
    blurb: "Two approaching masses. Collision then projection.",
  },
] as const;

export type Morphology = (typeof MORPHOLOGIES)[number]["id"];

function makeRng(seed: number) {
  let s = seed >>> 0;
  return () => {
    s = (Math.imul(s, 1664525) + 1013904223) >>> 0;
    return s / 4294967296;
  };
}

function fillBall(
  pos: Float32Array,
  vel: Float32Array,
  start: number,
  count: number,
  cx: number,
  cy: number,
  cz: number,
  radius: number,
  rng: () => number,
  vx = 0,
  vy = 0,
  vz = 0,
  inner = 0,
) {
  let written = 0;
  let guard = 0;
  const r2 = radius * radius;
  const i2 = inner * inner;
  while (written < count && guard < count * 40) {
    guard += 1;
    const x = (rng() * 2 - 1) * radius;
    const y = (rng() * 2 - 1) * radius;
    const z = (rng() * 2 - 1) * radius;
    const d2 = x * x + y * y + z * z;
    if (d2 > r2 || d2 < i2) continue;
    const i = start + written;
    pos[i * 3] = cx + x;
    pos[i * 3 + 1] = cy + y;
    pos[i * 3 + 2] = cz + z;
    vel[i * 3] = vx + (rng() - 0.5) * 0.02;
    vel[i * 3 + 1] = vy + (rng() - 0.5) * 0.02;
    vel[i * 3 + 2] = vz + (rng() - 0.5) * 0.02;
    written += 1;
  }
  while (written < count) {
    const u = rng() * Math.PI * 2;
    const v = Math.acos(rng() * 2 - 1);
    const rr = inner + (radius - inner) * Math.cbrt(rng());
    const i = start + written;
    pos[i * 3] = cx + rr * Math.sin(v) * Math.cos(u);
    pos[i * 3 + 1] = cy + rr * Math.cos(v);
    pos[i * 3 + 2] = cz + rr * Math.sin(v) * Math.sin(u);
    vel[i * 3] = vx;
    vel[i * 3 + 1] = vy;
    vel[i * 3 + 2] = vz;
    written += 1;
  }
}

function fillEllipsoid(
  pos: Float32Array,
  vel: Float32Array,
  count: number,
  cx: number,
  cy: number,
  cz: number,
  rx: number,
  ry: number,
  rz: number,
  rng: () => number,
) {
  let written = 0;
  let guard = 0;
  while (written < count && guard < count * 40) {
    guard += 1;
    const x = (rng() * 2 - 1) * rx;
    const y = (rng() * 2 - 1) * ry;
    const z = (rng() * 2 - 1) * rz;
    if ((x * x) / (rx * rx) + (y * y) / (ry * ry) + (z * z) / (rz * rz) > 1) {
      continue;
    }
    pos[written * 3] = cx + x;
    pos[written * 3 + 1] = cy + y;
    pos[written * 3 + 2] = cz + z;
    vel[written * 3] = (rng() - 0.5) * 0.02;
    vel[written * 3 + 1] = (rng() - 0.5) * 0.02;
    vel[written * 3 + 2] = (rng() - 0.5) * 0.02;
    written += 1;
  }
  while (written < count) {
    const u = rng() * Math.PI * 2;
    const v = Math.acos(rng() * 2 - 1);
    pos[written * 3] = cx + rx * Math.sin(v) * Math.cos(u) * 0.92;
    pos[written * 3 + 1] = cy + ry * Math.cos(v) * 0.92;
    pos[written * 3 + 2] = cz + rz * Math.sin(v) * Math.sin(u) * 0.92;
    vel[written * 3] = 0;
    vel[written * 3 + 1] = 0;
    vel[written * 3 + 2] = 0;
    written += 1;
  }
}

export function seedParticles(
  kind: Morphology,
  pos: Float32Array,
  vel: Float32Array,
  count: number,
) {
  const rng = makeRng(17 + kind.length * 97);
  vel.fill(0);
  switch (kind) {
    case "protoplast":
      fillBall(pos, vel, 0, count, 0.5, 0.5, 0.5, 0.22, rng);
      break;
    case "budding":
      fillBall(pos, vel, 0, count, 0.5, 0.5, 0.5, 0.2, rng);
      break;
    case "fission":
      fillEllipsoid(pos, vel, count, 0.5, 0.5, 0.5, 0.28, 0.16, 0.16, rng);
      break;
    case "polar":
      fillBall(pos, vel, 0, count, 0.5, 0.46, 0.5, 0.18, rng);
      break;
    case "blastula":
      fillBall(pos, vel, 0, count, 0.5, 0.5, 0.5, 0.24, rng, 0, 0, 0, 0.175);
      break;
    case "dual": {
      const half = Math.floor(count / 2);
      fillBall(pos, vel, 0, half, 0.32, 0.5, 0.5, 0.14, rng, 0.22, 0, 0);
      fillBall(
        pos,
        vel,
        half,
        count - half,
        0.68,
        0.5,
        0.5,
        0.14,
        rng,
        -0.22,
        0,
        0,
      );
      break;
    }
  }
}

function nodeCoord(i: number, n: number) {
  return n <= 1 ? 0.5 : i / (n - 1);
}

export function seedGrowth(
  kind: Morphology,
  growth: Float32Array,
  nx: number,
  ny: number,
  nz: number,
) {
  growth.fill(0);
  let n = 0;
  for (let k = 0; k < nz; k++) {
    const z = nodeCoord(k, nz);
    for (let j = 0; j < ny; j++) {
      const y = nodeCoord(j, ny);
      for (let i = 0; i < nx; i++) {
        const x = nodeCoord(i, nx);
        let g = 0;
        if (kind === "protoplast") {
          const dx = x - 0.5;
          const dy = y - 0.5;
          const dz = z - 0.5;
          const r2 = dx * dx + dy * dy + dz * dz;
          g = r2 < 0.08 ? 0.12 : 0;
        } else if (kind === "budding") {
          const dx = x - 0.66;
          const dy = y - 0.56;
          const dz = z - 0.5;
          g = 1.7 * Math.exp(-(dx * dx + dy * dy + dz * dz) / (2 * 0.055 * 0.055));
        } else if (kind === "fission") {
          const band = Math.exp(-((x - 0.5) * (x - 0.5)) / (2 * 0.042 * 0.042));
          const poles = Math.max(0, Math.abs(x - 0.5) - 0.1) * 4;
          g = -1.35 * band + 0.55 * Math.min(1, poles);
        } else if (kind === "polar") {
          const dx = x - 0.5;
          const dy = y - 0.72;
          const dz = z - 0.5;
          g = 1.85 * Math.exp(-(dx * dx + dy * dy + dz * dz) / (2 * 0.07 * 0.07));
        } else if (kind === "blastula") {
          const dx = x - 0.5;
          const dy = y - 0.5;
          const dz = z - 0.5;
          const r = Math.sqrt(dx * dx + dy * dy + dz * dz);
          g = Math.exp(-((r - 0.2) * (r - 0.2)) / (2 * 0.03 * 0.03)) * 0.7;
        } else {
          g = 0.18;
        }
        growth[n++] = g;
      }
    }
  }
}

File: python/morphogenesis/seeds.py
"""Seed morphologies. Computational only — no biological interpretation."""

from __future__ import annotations

import numpy as np

MORPHOLOGIES = (
    "protoplast",
    "budding",
    "fission",
    "polar",
    "blastula",
    "dual",
)


def _rng(seed: int) -> np.random.Generator:
    return np.random.default_rng(seed)


def _fill_ball(pos, vel, start, count, cx, cy, cz, radius, rng, vx=0.0, vy=0.0, vz=0.0, inner=0.0):
    written = 0
    guard = 0
    r2 = radius * radius
    i2 = inner * inner
    while written < count and guard < count * 40:
        guard += 1
        x, y, z = (rng.random(3) * 2 - 1) * radius
        d2 = float(x * x + y * y + z * z)
        if d2 > r2 or d2 < i2:
            continue
        i = start + written
        pos[i] = (cx + x, cy + y, cz + z)
        vel[i] = (
            vx + (rng.random() - 0.5) * 0.02,
            vy + (rng.random() - 0.5) * 0.02,
            vz + (rng.random() - 0.5) * 0.02,
        )
        written += 1
    while written < count:
        u = rng.random() * np.pi * 2
        v = np.arccos(rng.random() * 2 - 1)
        rr = inner + (radius - inner) * rng.random() ** (1 / 3)
        i = start + written
        pos[i] = (
            cx + rr * np.sin(v) * np.cos(u),
            cy + rr * np.cos(v),
            cz + rr * np.sin(v) * np.sin(u),
        )
        vel[i] = (vx, vy, vz)
        written += 1


def seed_particles(kind: str, count: int) -> tuple[np.ndarray, np.ndarray]:
    pos = np.zeros((count, 3), dtype=np.float64)
    vel = np.zeros((count, 3), dtype=np.float64)
    rng = _rng(17 + len(kind) * 97)
    if kind == "protoplast":
        _fill_ball(pos, vel, 0, count, 0.5, 0.5, 0.5, 0.22, rng)
    elif kind == "budding":
        _fill_ball(pos, vel, 0, count, 0.5, 0.5, 0.5, 0.20, rng)
    elif kind == "fission":
        written = 0
        guard = 0
        rx, ry, rz = 0.28, 0.16, 0.16
        while written < count and guard < count * 40:
            guard += 1
            x = (rng.random() * 2 - 1) * rx
            y = (rng.random() * 2 - 1) * ry
            z = (rng.random() * 2 - 1) * rz
            if (x * x) / (rx * rx) + (y * y) / (ry * ry) + (z * z) / (rz * rz) > 1:
                continue
            pos[written] = (0.5 + x, 0.5 + y, 0.5 + z)
            vel[written] = (
                (rng.random() - 0.5) * 0.02,
                (rng.random() - 0.5) * 0.02,
                (rng.random() - 0.5) * 0.02,
            )
            written += 1
        while written < count:
            u = rng.random() * np.pi * 2
            v = np.arccos(rng.random() * 2 - 1)
            pos[written] = (
                0.5 + rx * np.sin(v) * np.cos(u) * 0.92,
                0.5 + ry * np.cos(v) * 0.92,
                0.5 + rz * np.sin(v) * np.sin(u) * 0.92,
            )
            written += 1
    elif kind == "polar":
        _fill_ball(pos, vel, 0, count, 0.5, 0.46, 0.5, 0.18, rng)
    elif kind == "blastula":
        _fill_ball(pos, vel, 0, count, 0.5, 0.5, 0.5, 0.24, rng, inner=0.175)
    elif kind == "dual":
        half = count // 2
        _fill_ball(pos, vel, 0, half, 0.32, 0.5, 0.5, 0.14, rng, vx=0.22)
        _fill_ball(pos, vel, half, count - half, 0.68, 0.5, 0.5, 0.14, rng, vx=-0.22)
    else:
        raise ValueError(f"unknown morphology: {kind}")
    return pos, vel


def seed_growth(kind: str, nx: int, ny: int, nz: int) -> np.ndarray:
    g = np.zeros((nz, ny, nx), dtype=np.float64)
    zs = np.linspace(0.0, 1.0, nz) if nz > 1 else np.array([0.5])
    ys = np.linspace(0.0, 1.0, ny) if ny > 1 else np.array([0.5])
    xs = np.linspace(0.0, 1.0, nx) if nx > 1 else np.array([0.5])
    Z, Y, X = np.meshgrid(zs, ys, xs, indexing="ij")
    if kind == "protoplast":
        r2 = (X - 0.5) ** 2 + (Y - 0.5) ** 2 + (Z - 0.5) ** 2
        g = np.where(r2 < 0.08, 0.12, 0.0)
    elif kind == "budding":
        d2 = (X - 0.66) ** 2 + (Y - 0.56) ** 2 + (Z - 0.5) ** 2
        g = 1.7 * np.exp(-d2 / (2 * 0.055 * 0.055))
    elif kind == "fission":
        band = np.exp(-((X - 0.5) ** 2) / (2 * 0.042 * 0.042))
        poles = np.clip(np.abs(X - 0.5) - 0.1, 0, None) * 4
        g = -1.35 * band + 0.55 * np.minimum(1.0, poles)
    elif kind == "polar":
        d2 = (X - 0.5) ** 2 + (Y - 0.72) ** 2 + (Z - 0.5) ** 2
        g = 1.85 * np.exp(-d2 / (2 * 0.07 * 0.07))
    elif kind == "blastula":
        r = np.sqrt((X - 0.5) ** 2 + (Y - 0.5) ** 2 + (Z - 0.5) ** 2)
        g = np.exp(-((r - 0.2) ** 2) / (2 * 0.03 * 0.03)) * 0.7
    else:
        g = np.full_like(g, 0.18)
    return g
